internal/k8s/controllers/config_controller_test.go

```diff
@@ -0,0 +1,279 @@
+			Client:         fakeClient,
+			Logger:         log.NewNopLogger(),
+			Scheme:         scheme,
+			Namespace:      testNamespace,
+			ValidateConfig: config.DontValidate,
+			Handler:        mockHandler,
+			ForceReload:    mockForceReload,
+		}
+		req := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Namespace: testNamespace,
+			},
+		}
+
+		_, err = r.Reconcile(context.TODO(), req)
+		failedReconcile := err != nil
+
+		if test.expectReconcileFails != failedReconcile {
+			t.Errorf("%s: fail reconcile expected: %v, got: %v", test.desc, test.expectReconcileFails, failedReconcile)
+		}
+
+		if test.expectForceReloadCalled != calledForceReload {
+			t.Errorf("%s: call force reload expected: %v, got: %v", test.desc, test.expectForceReloadCalled, calledForceReload)
+		}
+	}
+}
+
+var (
+	testNamespace                  = "test-controller"
+	scheme                         = runtime.NewScheme()
+	configControllerValidResources = metallbcfg.ClusterResources{
+		Peers: []v1beta2.BGPPeer{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "peer1",
+					Namespace: testNamespace,
+				},
+				Spec: v1beta2.BGPPeerSpec{
+					MyASN:      42,
+					ASN:        142,
+					Address:    "1.2.3.4",
+					BFDProfile: "default",
+				},
+			},
+		},
+		BFDProfiles: []v1beta1.BFDProfile{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "default",
+					Namespace: testNamespace,
+				},
+			},
+		},
+		Pools: []v1beta1.IPAddressPool{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "pool1",
+					Namespace: testNamespace,
+				},
+				Spec: v1beta1.IPAddressPoolSpec{
+					Addresses: []string{
+						"10.20.0.0/16",
+					},
+				},
+			},
+		},
+		BGPAdvs: []v1beta1.BGPAdvertisement{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "adv1",
+					Namespace: testNamespace,
+				},
+			},
+		},
+		L2Advs: []v1beta1.L2Advertisement{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "l2adv1",
+					Namespace: testNamespace,
+				},
+			},
+		},
+		PasswordSecrets: map[string]corev1.Secret{
+			"bgpsecret": {Type: corev1.SecretTypeBasicAuth, ObjectMeta: metav1.ObjectMeta{Name: "bgpsecret", Namespace: testNamespace},
+				Data: map[string][]byte{"password": []byte([]byte("nopass"))}},
+		},
+		LegacyAddressPools: []v1beta1.AddressPool{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "legacypool1",
+					Namespace: testNamespace,
+				},
+				Spec: v1beta1.AddressPoolSpec{
+					Addresses: []string{
+						"10.21.0.0/16",
+					},
+					Protocol: "bgp",
+				},
+			},
+		},
+	}
+	configControllerInvalidResources = metallbcfg.ClusterResources{
+		Peers: []v1beta2.BGPPeer{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "peer1",
+					Namespace: testNamespace,
+				},
+			},
+		},
+	}
+)
+
+func newFakeClient(t *testing.T, resources metallbcfg.ClusterResources) client.WithWatch {
+	if err := v1beta1.AddToScheme(scheme); err != nil {
+		t.Fatalf("add to scheme failed: %s", err)
```

Comment:
maybe mention here which AddToScheme failed? they all return the same output so it might be hard to know which line failed

Reply:
done